simapp/migration_test.go

```diff
@@ -0,0 +1,167 @@
+		expCalled    int
+	}{
+		{
+			"cannot register migration for version 0",
+			"bank", 0,
+			true, "module migration versions should start at 1: invalid version", false, "", 0,
+		},
+		{
+			"throws error on RunMigrations if no migration registered for bank",
+			"", 1,
+			false, "", true, "no migrations found for module bank: not found", 0,
+		},
+		{
+			"can register and run migration handler for x/bank",
+			"bank", 1,
+			false, "", false, "", 1,
+		},
+		{
+			"cannot register migration handler for same module & forVersion",
+			"bank", 1,
+			true, "another migration for module bank and version 1 already exists: internal logic error", false, "", 0,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var err error
+
+			// Since it's very hard to test actual in-place store migrations in
+			// tests (due to the difficulty of maintaining multiple versions of a
+			// module), we're just testing here that the migration logic is
+			// called.
+			called := 0
+
+			if tc.moduleName != "" {
+				// Register migration for module from version `forVersion` to `forVersion+1`.
+				err = app.configurator.RegisterMigration(tc.moduleName, tc.forVersion, func(sdk.Context) error {
+					called++
+
+					return nil
+				})
+
+				if tc.expRegErr {
+					require.EqualError(t, err, tc.expRegErrMsg)
+
+					return
+				}
+			}
+			require.NoError(t, err)
+
+			// Run migrations only for bank. That's why we put the initial
+			// version for bank as 1, and for all other modules, we put as
+			// their latest ConsensusVersion.
+			_, err = app.mm.RunMigrations(
+				app.NewContext(true, tmproto.Header{Height: app.LastBlockHeight()}), app.configurator,
+				module.VersionMap{
+					"bank":         1,
+					"cert":         cert.AppModule{}.ConsensusVersion(),
+					"cvm":          cvm.AppModule{}.ConsensusVersion(),
+					"oracle":       oracle.AppModule{}.ConsensusVersion(),
+					"shield":       shield.AppModule{}.ConsensusVersion(),
+					"auth":         auth.AppModule{}.ConsensusVersion(),
+					"authz":        authzmodule.AppModule{}.ConsensusVersion(),
+					"staking":      staking.AppModule{}.ConsensusVersion(),
+					"mint":         mint.AppModule{}.ConsensusVersion(),
+					"distribution": distribution.AppModule{}.ConsensusVersion(),
+					"slashing":     slashing.AppModule{}.ConsensusVersion(),
+					"gov":          gov.AppModule{}.ConsensusVersion(),
+					"params":       sdkparams.AppModule{}.ConsensusVersion(),
+					"upgrade":      upgrade.AppModule{}.ConsensusVersion(),
+					"vesting":      vesting.AppModule{}.ConsensusVersion(),
+					"feegrant":     feegrantmodule.AppModule{}.ConsensusVersion(),
+					"evidence":     evidence.AppModule{}.ConsensusVersion(),
+					"crisis":       crisis.AppModule{}.ConsensusVersion(),
+					"genutil":      genutil.AppModule{}.ConsensusVersion(),
+					"capability":   capability.AppModule{}.ConsensusVersion(),
+					"transfer":     transfer.AppModule{}.ConsensusVersion(),
```

Comment:
I think gov module needs to be 1? since we have to migrate it to remove txhashes